### ISSUE TITLE
[select] support `rootState` select

### ISSUE THREAD

**Issue by leesiongchan** [rematch/rematch#270]
In my case I have to connect to other states to complete my selector, is there any reasons why do we limit it to `model-level` only?

**Comment by ShMcK** [comment]
Also noted in #276. 

@blairbodnar what are your thoughts on this one?

**Comment by jackwilsdon** [comment]
Maybe something like this would work?

```JavaScript
const model = {
  selectors: {
    getData: (localState, rootState) => ...
  }
};
```

**Comment by blairbodnar** [comment]
Hey guys! 

I think that we wrote the select plugin with the idea that it might be useful if you have _big_ models...

Let's say we have 3 counters. Maybe you choose to house all the logic in one model called `counters`. The select plugin might be useful here:

#### A bigger model:
##### models.js
```js
const counters = {
  state: {
    a: 1,
    b: 2,
    c: 3,
  },
  select: {
    sum: state => state.a + state.b + state.c
  }
}
```

#### Smaller models:
However, if you wrote it with 3 _smaller_ models, the select plugin wouldn't be much use...
##### models.js
```js
export const a = {
  state: 1
}

export const b = {
  state: 2
}

export const c = {
  state: 3
}
```
... but you could just write some selectors outside the model to bring your state together:
##### selectors.js
```js
export const sum = state => state.a + state.b + state.c
```

That being said...

@leesiongchan I would just write your own selectors outside of the models like the 2nd example above.
 (and not use the plugin). 😊

@jackwilsdon That's a good suggestion... but hmm... I'm thinking it might be best to keep it isolated and just write selectors outside of the model.

What do you guys think?

**Comment by will-stone** [comment]
I feel that the select plugin could function without the need to be passed any parameters, i.e it’s always passed local state

**Comment by jackwilsdon** [comment]
That makes sense, @blairbodnar. Maybe the `select` plugin can also offer a separate set of "global" selectors in that case?

```JavaScript
const store = init({
  models: {
    first: { state: 5 },
    second: { state: 10 }
  },
  plugins: [ select ],
  select: {
    total: ({ first, second }) => first + second
  }
});

store.select.total(store.getState()); // 15
```

**Comment by will-stone** [comment]
@jackwilsdon I think getState is being deprecated.

**Comment by leesiongchan** [comment]
Keeping it isolated sounds good to me.

**Comment by jackwilsdon** [comment]
@will-stone Looking at the code, that only seems to be the root `getState` method called like so;

```

**Comment by will-stone** [comment]
@jackwilsdon ah, I see. Thanks for clearing that up for me 😄

**Comment by ShMcK** [comment]
I have a proposal which may sound terrible:

- rename "selectors" to "computed", computed will only reference modelState
- change "selectors" to use "rootState"

**Comment by nWidart** [comment]
Maybe concerning the names, we could use inspiration from vuex (redux equivalent in vuejs) : https://vuex.vuejs.org/en/core-concepts.html

**Comment by will-stone** [comment]
I was looking into making a "computed" plugin, the other day, before realising that selectors weren't too far off. However, I'd definitely vote for computed, it'll mean these values can be accessed without having to import `select` into `mapStateToProps` files (I presume?).

**Comment by jackwilsdon** [comment]
Just threw together a plugin which uses `Object.defineProperty` to provide computed properties (from @will-stone's idea).

```JavaScript
const applyComputators = (state, computators) =>
  Object.keys(computators).reduce(
    (newState, modelName) => ({
      ...newState,
      [modelName]: Object.keys(computators[modelName]).reduce(
        (newModelState, computatorName) => {
          const computator = computators[modelName][computatorName];

          return Object.defineProperty(newModelState, computatorName, {
            enumerable: true,
            get: () =>
              computators[modelName][computatorName](newState[modelName]),
          });
        },
        { ...newState[modelName] },
      ),
    }),
    state,
  );

const computedPlugin = {
  init: exposed => {
    let computators = {};

    return {
      onStoreCreated(store) {
        const getState = store.getState;

        store.getState = () => applyComputators(getState(), computators);
      },
      onModel(model) {
        if (!model.computed) {
          return;
        }

        computators[model.name] = model.computed;
      },
    };
  },
};

export default computedPlugin;
```

There are a few improvements that could be made, namely caching the computor's result until the next dispatch happens (we could use middleware to clear the cache).

**Comment by blairbodnar** [comment]
Awesome ideas @jackwilsdon and @will-stone !

@jackwilsdon I like how you intercept `getState` - cool idea! 😀I did some playing around with that and here's two (crazy?) ideas I came up with:

### Global Selector Plugin Idea:
#### Working example here: https://codesandbox.io/s/pyk4lpv627

Whenever `store.getState()` is called, the selectors are computed and put in the store under the `stateKey`. I supposed this could lead to performance problems if there are a lot of expensive selectors, but maybe it wouldn't be that bad... hmm.

<img width="960" alt="selectors-plugin" src="https://user-images.githubusercontent.com/991976/37456660-2346aef6-27fd-11e8-9ffc-85d0898bc16d.png">

### Computed Plugin Idea
#### Working example here: https://codesandbox.io/s/p7yv0zoy4q

Same idea as @jackwilsdon above, but stored in a different `stateKey` than the model. (This would solve the problem of not being able to use primitives as you describe above. But it doesn't allow one computed property to access another computed property.  Could probably figure out a way to get that to work, but brain is dead now, need sleep, haha). I was also playing around with passing the rootState to the computed properties.

<img width="1046" alt="computed-plugin" src="https://user-images.githubusercontent.com/991976/37456835-903e1b70-27fd-11e8-804a-1622f2f584de.png">

**Comment by jackwilsdon** [comment]
Good ideas @blairbodnar!

With the global selector plugin, would it not be better to leave them as functions bound to the state so they can be called whenever (and with arguments)? Calling them automatically is almost like having a global computed state :stuck_out_tongue:

**Comment by titouancreach** [comment]
I'm taking part to this discussion, disclaimer, this will maybe be off-topic. 
I'm almost ended up refactoring my entire code base. I really like how do you introduce model to separate local state to global state. As I know redux, it I understood how do you designed rematch (maybe). But maybe we can go further with these model. Why if we drop the mapDispatchToProps and mapStateToProps to adopt mapModelToProps:

```javascript
const mapModelToProps = ({state, blogPost, authors, comments}, {currentBlogPostId}) => ({
  addComment: (comment) => comments.add(currentBlogPostId, comment),
  comments: comments.select.ownedBy(blogPost)
});
```
or
```javascript
const mapModelToProps = ({state, blogPost, authors, comments}, {currentBlogPostId}) => ({
  addComment: (comment) => blogPost.add(currentBlogPostId, comment),
  comments: blogPost.select.comments(blogPost)
});
```

**Comment by titouancreach** [comment]
One of the possible solution would be to get the store instance in the selectors (and reducers/effects) instead of the state, and bind select method to the store.

It could be used as follow:

```javascript
selectors: {
  mySelectorFoo: (store, myParam) {
    // const localState = store.getState.myModel() // if getState would work exactly as dispatch :)
    // const fullState = store.getState()
    // use another model selector or another state
    // const selected: store.select.MyOtherModel.myOtherSelector()
  }
}
```

The question we could ask if, but in mapStateToProps how could we select ?

Hmmmm, we can drop this and use mapStoreToProps. You made the things so easy, we don't really need to separate action from state. With selector, state needs no modification, with dispatch, we don't need to boundActionsCreators.

```javascript
const mapStoreToProps = ({state, select, dispatch}) => ({
  comments: select.comments.getAll(),
  addComment: dispatch.comments.add
});
  
```
We can even drop the state since we have selectors :D

It's currently a draft in my head, but I like this idea :+1:

**Comment by mBeierl** [comment]
I'd have to agree with @blairbodnar's 1st response on this issue. A model's selectors should only select from the model's state. Anything else would be misleading.

I think this is confusing if you worked with redux & reselect before, especially if you migrate to rematch, as I did. I had to rewrite my "low level" selectors like `const getFonts = state => state.fonts;` becomes `const getFonts = state => state;`.
But when starting out with rematch, structuring your state by model actually makes a lot of sense. I'd even consider it a good practice to follow.

Any selectors that would need to access other models should really live somewhere else, and not be part of a model's selectors.

**Comment by blairbodnar** [comment]
Great discussions everyone around selectors and computed properties. 

 I'll close this off for now in favour of #310 as it addresses the original topic of this thread. Released as `@rematch/select` **0.3.0** (Thanks for the PR @taranda).  

[`0.3.0` gives the plugin a `sliceState` config option which allows you to pass the rootState to model selectors if desired.](https://github.com/rematch/rematch/blob/master/plugins/select/README.md#configuration-options)

Feel free to make a PR or new issue for your selector / computed plugin ideas. 😎😄

Thanks!
